### PR TITLE
Improvements

### PR DESCRIPTION
**Pull request by zainab-ali**

### CODE REVIEW

File: core/src/main/scala/iliad/Bounded.scala [hunk 1]
```diff
@@ -7,3 +7,12 @@ trait Bounded[A] {
...
 }
+
+object Bounded {
+def index[A](n: Int)(implicit B: Bounded[A], E: Enum[A]) = {
+    @annotation.tailrec
+    def go(a: A, n: Int): A = if(n <= 0) a else go(Enum[A].succ(a), n - 1)
+    go(Bounded[A].MinValue, n)
+
+  }
+}
```
Comment (zainab-ali) [inline core/src/main/scala/iliad/Bounded.scala:24]: Should this method be somewhere else?

File: core/src/main/scala/iliad/gl/cachedgl.scala [hunk 2]
```diff
@@ -139,23 +139,122 @@ object CachedGL {
...
+                  ts: Map[String, Texture.Constructor]): DSL[String Xor Unit] =
+    p.textureUniforms(ts) //Xor[String, List[Uniform]]
+      .map(
+          us =>
+            us.map(set) //Xor[String, List[DSL[Xor[String, Unit]]]]
+              .sequence
+              .map(_.sequence.map(_ => ()))) //Xor[DSL[Xor[String, Unit]]]
+      .sequence
+      .map(_.flatMap(identity)) //DSL[Xor[String, Unit]]
+
```
Comment (zainab-ali) [inline core/src/main/scala/iliad/gl/cachedgl.scala:231]: This should be written better

File: core/src/main/scala/iliad/Bounded.scala (hunk 1 above)
Comment (yilinwei) [inline core/src/main/scala/iliad/Bounded.scala:24]: Can change to `A : Bounded : Enum`. It's good where it is, though would be worth returning an `Option` if it goes above max value.

File: core/src/main/scala/iliad/gl/cachedgl.scala [hunk 3]
```diff
@@ -139,23 +139,122 @@ object CachedGL {
...
+    }
+
+  private def attachment(a: Framebuffer.AttachmentConstructor)
+    : DSL[String Xor Framebuffer.AttachmentLoaded] = a match {
+    case t: Texture.Constructor =>
+      Cached
+        .ensure(Cached.get(t), s"Texture missing $t")
+        .freekF[CachedGL]
+        .map(identity)
+    case r: Renderbuffer.Constructor =>
```
Comment (yilinwei) [inline core/src/main/scala/iliad/gl/cachedgl.scala:154]: What is this for?

File: core/src/main/scala/iliad/gl/cachedgl.scala [hunk 4]
```diff
@@ -139,23 +139,122 @@ object CachedGL {
...
+          fl <- xort(Load(f, as).freekF[CachedGL])
+          _ <- xort(Cached.put(fl).freekF[CachedGL])
+        } yield ()).value
+    }
+
   def clear(bitMask: ChannelBitMask): DSL[Unit] =
     Draw.clear(bitMask).freekF[CachedGL]
 
   private def doIfNot(f: Current.DSL[Boolean])(g: DSL[Unit]): DSL[Unit] =
     f.freekF[CachedGL] flatMap (b => if (b) Free.pure(()) else g)
```
Comment (yilinwei) [inline core/src/main/scala/iliad/gl/cachedgl.scala:182]: Let `g` be an `Eval` or lazy. Can we also call this something different? I think `ensure` is the standard ish name.

File: core/src/main/scala/iliad/gl/cachedgl.scala (hunk 2 above)
Comment (yilinwei) [inline core/src/main/scala/iliad/gl/cachedgl.scala:231]: `.map.sequence` can be turned into `.traverse`.

File: core/src/main/scala/iliad/gl/data.scala [hunk 5]
```diff
@@ -166,20 +191,103 @@ object ElementData {
...
+      extends Framebuffer.AttachmentLoaded
+}
+
+
+
+object Framebuffer {
+  sealed trait Constructor {
+    def attachments: List[(FramebufferAttachment, AttachmentConstructor)]
+    def textures = attachments flatMap {
+      case (_, t: Texture.Constructor) => Some(t)
```
Comment (zainab-ali) [inline core/src/main/scala/iliad/gl/data.scala]: replace this with a `filterType` method

File: core/src/main/scala/iliad/gl/opengl.scala [hunk 6]
```diff
@@ -178,6 +182,104 @@ object GL {
...
+    }.sequence.map(_ => ())
+
+  private def makeFramebuffer(
+      as: List[(FramebufferAttachment, Framebuffer.AttachmentLoaded)],
+      id: Int)(f: Texture.Loaded => Int): DSL[Unit] =
+    for {
+      _ <- bindFramebuffer(id)
+      _ <- bindAttachments(as)(f)
+      _ <- GLDrawBuffers(as.map {
+            case (b: ColorBuffer, _) => Some(b)
```
Comment (zainab-ali) [inline core/src/main/scala/iliad/gl/opengl.scala]: replace this with a `filterType` method

File: core/src/main/scala/iliad/gl/cachedgl.scala [hunk 7]
```diff
@@ -139,23 +139,122 @@ object CachedGL {
...
+        for {
+          tl <- Load(t, d).freekF[CachedGL]
+          _ <- Cached.put(tl).freekF[CachedGL]
+        } yield tl
+    }
+
+  def renderbuffer(r: Renderbuffer.Constructor): DSL[Renderbuffer.Loaded] =
+    Cached.get(r).freekF[CachedGL] flatMap {
+      case Some(rl) => Free.pure(rl)
+      case None =>
```
Comment (zainab-ali) [inline core/src/main/scala/iliad/gl/cachedgl.scala]: This is done so often - it should have a helper method

File: core/src/main/scala/iliad/gl/cachedgl.scala (hunk 3 above)
Comment (zainab-ali) [inline core/src/main/scala/iliad/gl/cachedgl.scala:154]: This should be `widen`

**Pull request merged**